src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -78,11 +78,20 @@ public abstract class Compilation : ICompilation
 
         public virtual MethodIL GetMethodIL(MethodDesc method)
         {
-            // Flush the cache when it grows too big
-            if (_methodILCache.Count > 1000)
-                _methodILCache = new ILCache(_methodILCache.ILProvider);
+            try
```

Comment:
What failures does this help with? I would expect IL generation failure not to be a common occurrence except when working on a new feature (in which case we're usually running under a debugger anyway).

IL compilation happens elsewhere and we already have infrastructure for reporting that;

https://github.com/dotnet/corert/blob/a8271be78880b2037cc884b86246cc966d9425eb/src/ILCompiler/src/Program.cs#L808-L828

Reply:
See #8106 for this specific example.

Reply:
I see - it seems like we just need to hook up the scanner into the `CodeGenerationFailedException` infrastructure.

Instead of this, can you add one more catch clause to the try block in `ILScanner.CompileSingleMethod` - one that would catch all exceptions and rethrow them as `CodeGenerationFailedException`? I think that should achieve what you need as well (if not, we can tweak it more).

Reply:
@MichalStrehovsky can you take a look again. My hands are itching to play with Linux builds.